### PR TITLE
dont show remote and email options if we have an exact match for local user email

### PR DESCRIPTION
**Pull request by icewind1991**
Signed-off-by: Robin Appelman <robin@icewind.nl>

### CODE REVIEW

**Comment by wiswedel** [comment]
ok: guest users aren't shown for sharebymail anymore
not ok: also valid federated share targets aren't assumed anymore

before:
image

image


after:
image

image

File: lib/private/Collaboration/Collaborators/Search.php [hunk 1]
```diff
@@ -85,6 +85,13 @@ public function search($search, array $shareTypes, $lookup, $limit, $offset) {
 			$searchResult->unsetResult($emailType);
 		}
 
+		// if we have an exact local user match, there is no need to show the remote and email matches
+		$userType = new SearchResultType('users');
+		if($searchResult->hasExactIdMatch($userType)) {
+			$searchResult->unsetResult($remoteType);
```
Comment (wiswedel) [inline lib/private/Collaboration/Collaborators/Search.php]: This one is too much

**Comment by icewind1991** [comment]
imo, if a local user with a given email exists than it's highly unlikely that that email is also a fed share target that the user wants to share with. It already hides the fed. share option if the email is found in the user's address book

**Comment by wiswedel** [comment]
> It already hides the fed. share option if the email is found in the user's address book

Fair enough. Didn't know that

> if a local user with a given email exists than it's highly unlikely that that email is also a fed share target that the user wants to share with.

Tricky to assume things like that. But I'm not going to open a general discussion about this. So let's go for your solution.

**Review by wiswedel — APPROVED** [review]

**Comment by wiswedel** [comment]
@rullzer @blizzz Can you please add your review? We need to get this on the road. Thx.

**Comment by MorrisJobke** [comment]
I just tested again and for email addresses with exact match this is already the case since 15 at least. See also https://github.com/nextcloud/server/issues/15665#issuecomment-513802475

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by MorrisJobke** [comment]
Those are the tests that fail:

https://github.com/nextcloud/server/blob/03f1fef1601c21a2c387a20db9e46b21273c15c9/tests/lib/Collaboration/Collaborators/UserPluginTest.php#L244-L358

All of them do things like this:

```
				[
					['abc', 'test', 2, 0, ['test1' => 'Test One']],
					['xyz', 'test', 2, 0, []],
				],
```

as the user results instead of 

```

				[
					$this->getUserMock('test0', 'Test'),
					$this->getUserMock('test1', 'Test One'),
					$this->getUserMock('test2', 'Test Two'),
				],
```

Thus I have no idea what the initial thoughts where behind this.

They are coming from somewhere here: https://github.com/nextcloud/server/blame/dd9e191d373217b2a07e4ac5b2cc294c0a6227a1/apps/files_sharing/tests/Controller/ShareesAPIControllerTest.php#L1272

@nickvergessen @icewind1991 Any idea how to "fix" them?

**Comment by nickvergessen** [comment]
Seems like the tests where just not updated. It should return getUserMocks there as well.

File: lib/private/Collaboration/Collaborators/UserPlugin.php [hunk 2]
```diff
@@ -71,7 +71,7 @@ public function search($search, $limit, $offset, ISearchResult $searchResult) {
 			foreach ($userGroups as $userGroup) {
 				$usersTmp = $this->groupManager->displayNamesInGroup($userGroup, $search, $limit, $offset);
 				foreach ($usersTmp as $uid => $userDisplayName) {
-					$users[$uid] = $userDisplayName;
+					$users[$uid] = $this->userManager->get($uid);
```
Comment (nickvergessen) [inline lib/private/Collaboration/Collaborators/UserPlugin.php]: This needs adjustments in the tests.
It returns tones of `null` which of course later on dont have a `getDisplayName()` function, etc.

**Comment by skjnldsv** [comment]
Bump?

**Comment by nickvergessen** [comment]
Rebased and fixed the tests.
Let's get this in

**Comment by blizzz** [comment]
another rebase, as some failing tests are fixed on master meanwhile

**Comment by icewind1991** [comment]
rebased and fixed tests

**Comment by skjnldsv** [comment]
Integration tests are broken

**Comment by skjnldsv** [comment]
@danxuliu can you have a look, I'm really not familiar with the integration tests :(

**Comment by danxuliu** [comment]
> @danxuliu can you have a look, I'm really not familiar with the integration tests :(

The sharees integration test failures are legit. The problem is that now `$userGroup->searchDisplayName` is used, and [it calls `$backend->usersInGroup`](https://github.com/nextcloud/server/blob/caff1023ea72bb2ea94130e18a2a6e2ccf819e5f/lib/private/Group/Group.php#L311), which performs a [case sensitive search](https://github.com/nextcloud/server/blob/734c62bee0f46652285f782761317889d1ef5c7d/lib/private/Group/Database.php#L344). As the search string in the tests is _sharee_ [the user _Sharee1_ now is not found](https://github.com/nextcloud/server/blob/ea6f423e2c8e50cf1357a0e2182dc4c9a9bf981e/build/integration/sharees_features/sharees.feature#L68).

**Comment by skjnldsv** [comment]
Yep, needed for 18! 
/backport to stable18

**Comment by nickvergessen** [comment]
This breaks searching for users by displaynames when `shareapi_only_share_with_group_members` is on see https://github.com/nextcloud/server/issues/21451

**Comment by DanScharon** [comment]
unfortunately this doesn't work if the mail address is pasted with preceding or following whitespace(s). 
We have users who do that and hit the first result (federated share) which fails, but they very often don't see the error message.

**Pull request merged**